include/RAJA/policy/cuda/MemUtils_CUDA.hpp

```diff
@@ -299,184 +299,207 @@ cudaDeviceProp& device_prop()
-      uninitialized_size_t, uninitialized, uninitialized, uninitialized};
-
-  if ( data.prev_shmem_size  != shmem_size ||
-       data.prev_num_threads != num_threads ) {
-
-    cudaErrchk(cudaOccupancyMaxActiveBlocksPerMultiprocessor(
-    &data.max_blocks, func, num_threads, shmem_size));
+  static thread_local CudaOccMaxBlocksData data {
+      std::numeric_limits<size_t>::max(),
+      -1,
+      cuda::device_prop().multiProcessorCount,
+      cuda::device_prop().maxThreadsPerMultiProcessor,
+      -1 };
 
-    if (data.multiProcessorCount == uninitialized) {
+  if ( data.func_dynamic_shmem_per_block != func_dynamic_shmem_per_block ||
+       data.func_threads_per_block != func_threads_per_block ) {
 
-      data.multiProcessorCount = cuda::device_prop().multiProcessorCount;
+    data.func_dynamic_shmem_per_block = func_dynamic_shmem_per_block;
+    data.func_threads_per_block = func_threads_per_block;
 
-    }
-
-    data.max_blocks *= data.multiProcessorCount;
-
-    data.prev_shmem_size  = shmem_size;
-    data.prev_num_threads = num_threads;
+    cudaErrchk(cudaOccupancyMaxActiveBlocksPerMultiprocessor(
+    &data.func_max_blocks_per_sm, func, func_threads_per_block, func_dynamic_shmem_per_block));
 
   }
 
-  max_blocks = data.max_blocks;
-
+  return data;
 }
 
-struct CudaOccupancyDefaults
+/*!
+ ******************************************************************************
+ *
+ * \brief  Cuda Concretizer Implementation.
+ *
+ * \tparam IdxT Index type to use for integer calculations.
+ * \tparam Concretizer Class the determines the max number of blocks to use when
+ *         fitting for the device.
+ * \tparam UniqueMarker A type that is unique to each global function, used to
+ *         help cache the occupancy data for that global function.
+ *
+ ******************************************************************************
+ */
+template < typename IdxT, typename Concretizer, typename UniqueMarker>
+struct ConcretizerImpl
 {
-  CudaOccupancyDefaults(const void* RAJA_UNUSED_ARG(func))
+  ConcretizerImpl(const void* func, size_t func_dynamic_shmem_per_block, IdxT len)
+    : m_func(func)
+    , m_func_dynamic_shmem_per_block(func_dynamic_shmem_per_block)
+    , m_len(len)
   { }
 
-  template < typename IdxT >
-  inline auto get_max_grid_size(size_t RAJA_UNUSED_ARG(dynamic_shmem_size),
-                                IdxT RAJA_UNUSED_ARG(block_size)) const
+  // Get the maximum block size
+  IdxT get_max_block_size() const
   {
-    return std::numeric_limits<IdxT>::max();
+    auto data = cuda_occupancy_max_blocks_threads<UniqueMarker>(
+        m_func, m_func_dynamic_shmem_per_block);
+    IdxT func_max_threads_per_block = data.func_max_threads_per_block;
+    return func_max_threads_per_block;
   }
 
-  template < typename IdxT = cuda_dim_member_t >
-  inline auto get_max_block_size_and_grid_size(size_t RAJA_UNUSED_ARG(dynamic_shmem_size)) const
+  // Get a block size that combined with the given grid size is large enough
```

Comment:
Comment is confusing. Where is the "given grid size"? Where is "len" specified?

Reply:
It might make sense to split this into a class for the for the `fit_len` versions of the methods and one of the `fit_device` methods. What do you think?

Reply:
That's your call. I think making things clear in the comments is sufficient.